Issue/247 duplicate resources

hhund opened the pull request:
* Adds check to make sure StructureDefinitions define a base definition
* Fixes transaction bundle error message
* Code cleanup
* Unique constraints implemented as constraint trigger run after insert
* Constraint trigger functions use postgres advisory transaction locks to ensure uniqueness checks are not executed in parallel
* Transaction isolation level of insert/update operations changed from repeatable read to read committed, enabling dirty reads needed to allow constraint triggers to see inserts/updates executed by parallel running transactions
* New integration test to validate parallel create operations via transaction and batch bundles as well as direct POSTs
* Delete queries to cleanup duplicate resources from production DBs
* Modified Task authorization rule to enforce unique draft Task resources

closes #247

hhund committed:
fix: different organization roles need to be different codes not codings

hhund committed:
added check to make sure StructureDefinition defines a base definition

check needed to avoid NPE in HAPI code

hhund committed:
transaction bundle error message fix

hhund committed:
added "throws WebApplicationException" to improve method signatures

hhund committed:
code cleanup

hhund committed:
database constraint trigger to ensure resource unique criteria

* Unique constraints implemented as constraint trigger run after insert
* Constraint trigger functions use postgres advisory transaction locks
to ensure uniqueness checks are not executed in parallel
* Transaction isolation level of insert/update operations changed from
repeatable read to read committed, enabling dirty reads needed to allow
constraint triggers to see inserts/updates executed by parallel running
transactions
* New integration test to validate parallel create operations via
transaction and batch bundles as well as direct POSTs

hhund committed:
added missing connection parameter

This does not change any behavior as currently no delete authorization
rule is dependent on the DB state.

hhund committed:
code cleanup, removed empty line

hhund committed:
test fix

hhund committed:
fixed db lock value and exception message typo

hhund committed:
delete queries to remove duplicate resources

wetret commented on dsf-fhir/dsf-fhir-server/src/main/java/dev/dsf/fhir/help/ResponseGenerator.java:
```diff
@@ -381,6 +381,23 @@ public Response multipleExists(String resourceTypeName, String ifNoneExistsHeade
 		return Response.status(Status.PRECONDITION_FAILED).entity(outcome).build();
 	}
 
+	public Response dupicateResourceExists()
```
I think there is a typo in the method name.

hhund replied:
fixed via 7c8ce81

hhund committed:
debug/audit log typo fixes

hhund committed:
improved audit log message

log message improved by early wrapping of PSQLException

hhund committed:
improved unique constraints

contains operators @> and <@ only work in one direction, added missing
directions to complete intersection checks

hhund committed:
fixed typos, added missing tests, improved test constants

hhund committed:
unique draft Task rules, db constraint, duplicate delete query and tests

hhund committed:
typo fix
